### ISSUE TITLE
no-param-reassign with {"props": true} should catch uses of Object.assign

### ISSUE THREAD

**Issue by dmose** [eslint/eslint#7570]
**Tell us about your environment**

* **ESLint Version:**
3.9.1

* **Node Version:**
6.9.1

* **npm Version:**
3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

The only relevant part:
```
    "no-param-reassign": [2, {"props": true}],
```

**What did you do? Please include the actual source code causing the issue.**

```js
function overwriteParamUsingObjectAssign(obj) {
  Object.assign(obj, {foo: 1});
}
```

**What did you expect to happen?**
An error should be triggered, since properties of obj are being overwritten using Object.assign.

**What actually happened? Please include the actual, raw output from ESLint.**
No error was triggered.

**Comment by kaicataldo** [comment]
I'm not sure I agree with this. While I personally generally prefer not to mutate objects by reference in function calls, it's a legitimate use case.

**Comment by ljharb** [comment]
When this rule is enabled, the current solution is then to use a disable comment.

I think it's simply an oversight that `Object.assign` with a non-object-literal as its first parameter isn't recognized by this rule as "mutation".

**Comment by not-an-aardvark** [comment]
@kaicataldo We already have a `props: true` option that disallows something like this:

```js
/* eslint no-param-reassign: [error, {props: true}] */
function foo(arg) {
  arg.baz = 5;
}
```

I think if someone sets `props: true`, they have explicitly opted into not mutating objects that are passed by reference, so I think this would be a good change. For example, sometimes I accidentally do this:

```js
function addOkProp(arg) {
  // return an object with all of `foo`'s properties, as well as {ok: true}, but don't modify foo
  return Object.assign(arg, {ok: true}); // oops, accidentally mutated foo.
}
```

However, it seems like it would also be a breaking change unless we can figure out a good way to opt into it.

**Comment by platinumazure** [comment]
I'm :-1: to the proposal as-is because someone could make the same case for something like `lodash.assign`.

What I might like to see is something similar to `no-restricted-properties`, but which only triggers when an enclosing function parameter (or other specific sort of argument) is passed into a function call. So that way, a user could flag `Object.assign`, `lodash.assign`, etc., as needed. But that would be hard to fit into core.

**Comment by kaicataldo** [comment]
Ah, my misunderstanding - sorry about that.

**Comment by ljharb** [comment]
@platinumazure a rule that restricted `Object.assign` usage with a non-object-literal as the first argument would cover this too, and could be configured to expand to a user-supplied list of identifiers?

**Comment by not-an-aardvark** [comment]
There are cases where `Object.assign` is useful for mutating objects, though.

```js
class Foo {
  constructor(props) {
    Object.assign(this, props);
  }
}
```

Personally, I think it would be more useful if it disallowed mutating things with `Object.assign` only when the mutation is a side-effect, i.e. it's allowed if the parent is an `ExpressionStatement`.

**Comment by not-an-aardvark** [comment]
> What I might like to see is something similar to `no-restricted-properties`, but which only triggers when an enclosing function parameter (or other specific sort of argument) is passed into a function call. So that way, a user could flag `Object.assign`, `lodash.assign`, etc., as needed. But that would be hard to fit into core.

@platinumazure I see where you're coming from, but it seems like such a rule would have to be extremely customizable, to the extent that the user might be better off just writing a custom rule.

**Comment by platinumazure** [comment]
> @platinumazure I see where you're coming from, but it seems like such a rule would have to be extremely customizable, to the extent that the user might be better off just writing a custom rule.

Yes, hence my last sentence "But that would be hard to fit into core."

I'm having a hard time with prioritizing `Object.assign` for special treatment here, that's my main point.

**Comment by not-an-aardvark** [comment]
My opinion is that since `Object.assign` is a native language feature, it might warrant special treatment that we wouldn't give to userland functions such as `lodash.assign`.

**Comment by fasttime** [comment]
What about `Reflect.set(obj, "foo", 1)`? Should that also be forbidden?

**Comment by not-an-aardvark** [comment]
Since `Reflect.set` is a native language feature, I suppose that should also be forbidden.

That said, I suppose using `Reflect.set` on a parameter is common with `Proxy`:

```js
const foo = new Proxy({}, {set(target, property, value) {
  console.log(`set ${property} to ${value}`);
  return Reflect.set(target, property, value); // <-- ideally this shouldn't get reported
}});
```

**Comment by fasttime** [comment]
@not-an-aardvark Indeed. I guess `Object.defineProperty` and `Object.defineProperties` deserve a different treatment. But I'm sure there are quite a few other native ways to manipulate properties on an object.

**Comment by not-an-aardvark** [comment]
Hmm, that's a good point. I think I'm going to backtrack a bit.

If someone really wants to mutate an object, they will find some way to get around ESLint and mutate it. If the user uses `Reflect.set` or `Object.defineProperty`, I think their intentions are very clear, so maybe we shouldn't report that case. On the other hand, it's fairly easy to *accidentally* mutate something with `Object.assign`, so it's still worth guarding against that.

**Comment by fasttime** [comment]
@not-an-aardvark Well, that sounds fair to me.

Just to summarize:

```js
obj["foo"] = 1;                                     // error
obj.foo = 1;                                        // error
Object.assign(obj, { foo: 1 });                     // error
Reflect.set(obj, "foo", 1);                         // ok
Object.defineProperty(obj, "foo", { value: 1 });    // ok
Object.defineProperties(obj, { foo: { value: 1 });  // ok
```

**Comment by dmose** [comment]
Seems reasonable to me too.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
